packages/date-picker/src/vaadin-date-picker-mixin.js

```diff
@@ -645,27 +659,50 @@ export const DatePickerMixin = (subclass) =>
       this._shouldKeepFocusRing = focused && this._keyboardActive;
     }
 
-    /** @private */
-    __dispatchChange() {
-      this.validate();
-      this.dispatchEvent(new CustomEvent('change', { bubbles: true }));
+    /**
+     * Depending on the nature of the value change that has occurred since
+     * the last commit attempt, triggers validation and fires an event:
+     *
+     * Value change             | Event
+     * :------------------------|:------------------
+     * empty => parsable        | change
+     * empty => unparsable      | unparsable-change
+     * parsable => empty        | change
+     * parsable => parsable     | change
+     * parsable => unparsable   | change
+     * unparsable => empty      | unparsable-change
+     * unparsable => parsable   | change
+     * unparsable => unparsable | unparsable-change
+     *
+     * @private
+     */
+    __commitValueChange() {
+      if (this.__committedValue !== this.value) {
+        this.validate();
+        this.dispatchEvent(new CustomEvent('change', { bubbles: true }));
+      } else if (this.__committedUnparsableValue !== this.__unparsableValue) {
+        this.validate();
+        this.dispatchEvent(new CustomEvent('unparsable-change'));
+      }
+
+      this.__committedValue = this.value;
+      this.__committedUnparsableValue = this.__unparsableValue;
     }
 
     /**
-     * Sets the given date as the value and fires a change event
-     * if the value has changed.
+     * Sets the given date as the value and commits it.
      *
      * @param {Date} date
      * @private
      */
     __commitDate(date) {
-      const prevValue = this.value;
-
+      // Prevent the value observer from treating the following value change
+      // as initiated programmatically by the developer, and therefore
+      // from automatically committing it without a change event.
+      this.__keepCommittedValue = true;
       this._selectedDate = date;
-
-      if (prevValue !== this.value) {
-        this.__dispatchChange();
-      }
+      this.__keepCommittedValue = false;
```

Comment:
Removing this line doesn't seem to affect any test.

Reply:
Good catch, thanks. Added a couple of more tests.

Reply:
Also added similar tests for NumberField and TimePicker.